com.hazelcast.map.impl.mapstore.offload.MapStoreOffloadingBouncingNodeTest.stress [HZ-2343]

_5.2.z_ (commit a43b66809cad8db3a54737d2026e4b3e6091370d)

Failure for variant `[writeBehindEnabled: true]`

Failed on openjdk-11: https://jenkins.hazelcast.com/job/Hazelcast-5.maintenance-ZuluJDK11/71/testReport/junit/com.hazelcast.map.impl.mapstore.offload/MapStoreOffloadingBouncingNodeTest/stress_writeBehindEnabled__true_/
<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.hazelcast.map.impl.mapstore.offload.MapStoreOffloadingBouncingNodeTest.lambda$stress$2(MapStoreOffloadingBouncingNodeTest.java:163)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1247)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1349)
	at com.hazelcast.map.impl.mapstore.offload.MapStoreOffloadingBouncingNodeTest.stress(MapStoreOffloadingBouncingNodeTest.java:160)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)

```
</details>



Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-5.maintenance-ZuluJDK11/71/

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/offload/MapStoreOffloadingBouncingNodeTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/offload/MapStoreOffloadingBouncingNodeTest.java
@@ -24,11 +24,10 @@ import com.hazelcast.internal.util.RuntimeAvailableProcessors;
 import com.hazelcast.map.EntryProcessor;
 import com.hazelcast.map.IMap;
 import com.hazelcast.map.MapStoreAdapter;
-import com.hazelcast.test.HazelcastParallelParametersRunnerFactory;
 import com.hazelcast.test.HazelcastParametrizedRunner;
+import com.hazelcast.test.HazelcastSerialParametersRunnerFactory;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
-import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
@@ -46,13 +45,15 @@ import java.util.concurrent.atomic.AtomicBoolean;
 
 import static com.hazelcast.map.IMapAccessors.getPendingOffloadedOpCount;
 import static com.hazelcast.spi.impl.operationservice.impl.OperationServiceAccessor.getAsyncOperationsCount;
+import static com.hazelcast.spi.impl.operationservice.impl.OperationServiceAccessor.toStringAsyncOperations;
+import static java.lang.String.format;
 import static java.util.Arrays.asList;
 import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.junit.Assert.assertEquals;
 
 @RunWith(HazelcastParametrizedRunner.class)
-@Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
-@Category({ParallelJVMTest.class, QuickTest.class})
+@Parameterized.UseParametersRunnerFactory(HazelcastSerialParametersRunnerFactory.class)
+@Category({QuickTest.class})
 public class MapStoreOffloadingBouncingNodeTest extends HazelcastTestSupport {
 
     @Parameterized.Parameters(name = "writeBehindEnabled: {0}")
@@ -67,52 +68,6 @@ public class MapStoreOffloadingBouncingNodeTest extends HazelcastTestSupport {
     public boolean writeBehindEnabled;
 
     private static final int TEST_RUN_SECONDS = 20;
-
-    protected Config getConfig(String mapName) {
-        Config config = smallInstanceConfigWithoutJetAndMetrics();
-        config.getMapConfig(mapName)
-                .setBackupCount(1)
-                .setAsyncBackupCount(0)
-                .setInMemoryFormat(getInMemoryFormat())
-                .getMapStoreConfig()
-                .setEnabled(true)
-                .setOffload(true)
-                .setWriteDelaySeconds(writeBehindEnabled ? 3 : 0)
-                .setImplementation(new MapStoreAdapter() {
-
-                    @Override
-                    public void store(Object key, Object value) {
-                        sleepRandomMillis();
-                        super.store(key, value);
-                    }
-
-                    @Override
-                    public void delete(Object key) {
-                        sleepRandomMillis();
-                        super.delete(key);
-                    }
-
-                    @Override
-                    public Object load(Object key) {
-                        sleepRandomMillis();
-                        return randomStringOrNull();
-                    }
-                });
-        return config;
-    }
-
-    protected InMemoryFormat getInMemoryFormat() {
-        return InMemoryFormat.BINARY;
-    }
-
-    private static Object randomStringOrNull() {
-        return RandomPicker.getInt(0, 10) < 2
-                ? null : randomString();
-    }
-
-    private static void sleepRandomMillis() {
-        sleepMillis(RandomPicker.getInt(0, 3));
-    }
 
     @Test(timeout = 5 * 60 * 1000)
     public void stress() throws InterruptedException {
@@ -160,10 +115,63 @@ public class MapStoreOffloadingBouncingNodeTest extends HazelcastTestSupport {
         assertTrueEventually(() -> {
             Collection<HazelcastInstance> instances = factory.getAllHazelcastInstances();
             for (HazelcastInstance node : instances) {
-                assertEquals(0, getAsyncOperationsCount(node));
-                assertEquals(0, getPendingOffloadedOpCount(node.getMap(mapName)));
+                if (node.getLifecycleService().isRunning()) {
+                    int asyncOperationsCount = getAsyncOperationsCount(node);
+                    int pendingOffloadedOpCount = getPendingOffloadedOpCount(node.getMap(mapName));
+
+                    assertEquals(format("Found asyncOperationsCount=%d {%s}",
+                            asyncOperationsCount, toStringAsyncOperations(node)), 0, asyncOperationsCount);
+                    assertEquals(format("Found pendingOffloadedOpCount=%d", pendingOffloadedOpCount),
+                            0, pendingOffloadedOpCount);
+                }
             }
         });
+    }
+
+    protected Config getConfig(String mapName) {
+        Config config = smallInstanceConfigWithoutJetAndMetrics();
+        config.getMapConfig(mapName)
+                .setBackupCount(1)
+                .setAsyncBackupCount(0)
+                .setInMemoryFormat(getInMemoryFormat())
+                .getMapStoreConfig()
+                .setEnabled(true)
+                .setOffload(true)
+                .setWriteDelaySeconds(writeBehindEnabled ? 3 : 0)
+                .setImplementation(new MapStoreAdapter() {
+
+                    @Override
+                    public void store(Object key, Object value) {
+                        sleepRandomMillis();
+                        super.store(key, value);
+                    }
+
+                    @Override
+                    public void delete(Object key) {
+                        sleepRandomMillis();
+                        super.delete(key);
+                    }
+
+                    @Override
+                    public Object load(Object key) {
+                        sleepRandomMillis();
+                        return randomStringOrNull();
+                    }
+                });
+        return config;
+    }
+
+    protected InMemoryFormat getInMemoryFormat() {
+        return InMemoryFormat.BINARY;
+    }
+
+    private static Object randomStringOrNull() {
+        return RandomPicker.getInt(0, 10) < 2
+                ? null : randomString();
+    }
+
+    private static void sleepRandomMillis() {
+        sleepMillis(RandomPicker.getInt(0, 3));
     }
 
     private enum OpType {

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/OperationServiceAccessor.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/OperationServiceAccessor.java
@@ -35,4 +35,10 @@ public final class OperationServiceAccessor {
         OperationServiceImpl operationService = getOperationService(instance);
         return operationService.asyncOperations.size();
     }
+
+    // only used for testing
+    public static String toStringAsyncOperations(HazelcastInstance instance) {
+        OperationServiceImpl operationService = getOperationService(instance);
+        return operationService.asyncOperations.toString();
+    }
 }

